File: sdk/go/hovel/module.go
package hovel

import "encoding/json"

// ModuleType is the kind of work a module performs. The daemon uses it to group
// modules and to reason about post-exploitation sessions.
type ModuleType string

const (
	// TypeSurvey gathers facts about a target without changing it.
	TypeSurvey ModuleType = "survey"
	// TypeExploit performs an offensive action that may open a session.
	TypeExploit ModuleType = "exploit"
	// TypePayloadProvider generates payloads for delivery by other modules.
	TypePayloadProvider ModuleType = "payload_provider"
)

// Requirement describes a single configuration field a module needs. It mirrors
// the daemon's requirement schema: chain-level requirements apply to a whole
// chain, target-level requirements apply per target.
type Requirement struct {
	// Key is the dotted configuration key, e.g. "target.host".
	Key string
	// Type is a value type the daemon validates against, e.g. "host", "port",
	// "bool", "string", "secret", "int", "url", "cidr".
	Type string
	// Required marks the field as mandatory before a throw can be planned.
	Required bool
	// Default is the value used when the operator does not set the field.
	Default string
	// Description is shown to operators configuring the module.
	Description string
	// Allowed optionally constrains the field to an enumerated set.
	Allowed []string
	// Secret hides the value in logs and transcripts.
	Secret bool
}

// ModuleContext is optional module-authored context for agent-facing catalog,
// planning, and step inspection paths. Hovel treats it as untrusted content
// with module provenance; omitting it has no effect on non-agent workflows.
type ModuleContext struct {
	Summary       string           `json:"summary,omitempty"`
	Keywords      []string         `json:"keywords,omitempty"`
	Platforms     []string         `json:"platforms,omitempty"`
	Targets       []string         `json:"targets,omitempty"`
	Capabilities  []string         `json:"capabilities,omitempty"`
	Preconditions []string         `json:"preconditions,omitempty"`
	SideEffects   []string         `json:"sideEffects,omitempty"`
	Cleanup       string           `json:"cleanup,omitempty"`
	Risk          RiskContext      `json:"risk,omitempty"`
	Examples      []ContextExample `json:"examples,omitempty"`
	AgentHints    []AgentHint      `json:"agentHints,omitempty"`
}

func (c ModuleContext) MarshalJSON() ([]byte, error) {
	type contextJSON struct {
		Summary       string           `json:"summary,omitempty"`
		Keywords      []string         `json:"keywords,omitempty"`
		Platforms     []string         `json:"platforms,omitempty"`
		Targets       []string         `json:"targets,omitempty"`
		Capabilities  []string         `json:"capabilities,omitempty"`
		Preconditions []string         `json:"preconditions,omitempty"`
		SideEffects   []string         `json:"sideEffects,omitempty"`
		Cleanup       string           `json:"cleanup,omitempty"`
		Risk          *RiskContext     `json:"risk,omitempty"`
		Examples      []ContextExample `json:"examples,omitempty"`
		AgentHints    []AgentHint      `json:"agentHints,omitempty"`
	}
	out := contextJSON{
		Summary:       c.Summary,
		Keywords:      c.Keywords,
		Platforms:     c.Platforms,
		Targets:       c.Targets,
		Capabilities:  c.Capabilities,
		Preconditions: c.Preconditions,
		SideEffects:   c.SideEffects,
		Cleanup:       c.Cleanup,
		Examples:      c.Examples,
		AgentHints:    c.AgentHints,
	}
	if riskContextPresent(c.Risk) {
		risk := c.Risk
		out.Risk = &risk
	}
	return json.Marshal(out)
}

type RiskContext struct {
	Level   string   `json:"level,omitempty"`
	Reasons []string `json:"reasons,omitempty"`
}

func riskContextPresent(risk RiskContext) bool {
	return risk.Level != "" || len(risk.Reasons) > 0
}

type ContextExample struct {
	Name        string            `json:"name,omitempty"`
	Description string            `json:"description,omitempty"`
	Modules     []string          `json:"modules,omitempty"`
	ChainConfig map[string]string `json:"chainConfig,omitempty"`
}

// Req constructs a required string Requirement; chain extra options as needed by
// mutating the returned value. It keeps example modules terse.
func Req(key, valueType, description string) Requirement {
	return Requirement{Key: key, Type: valueType, Required: true, Description: description}
}

// Info is the metadata a module reports during the handshake. Name, Version,
// and Type are required; Hovel treats this handshake as authoritative over any
// package-manifest hints.
type Info struct {
	Name             string
	Version          string
	Type             ModuleType
	Summary          string
	Description      string
	Tags             []string
	DiscoveryContext ModuleContext
}

// Schema is the configuration contract a module reports.
type Schema struct {
	ChainConfig     []Requirement
	TargetConfig    []Requirement
	Outputs         map[string]any
	PlanningContext ModuleContext
}

// Module is implemented by every Hovel module. Info and Schema must be cheap and
// side-effect free: the daemon calls them while building its catalog. Run does
// the actual work and is called once per (target) when the module is thrown.
type Module interface {
	Info() Info
	Schema() Schema
	Run(ctx *Context) (Result, error)
}

File: sdk/rust/hovel/src/module.rs
//! The module contract: metadata, configuration schema, and the [`Module`] trait.

use crate::context::Context;
use crate::mesh::{
    MeshBeacon, MeshBeaconRequest, MeshDescriptor, MeshDescribeRequest,
    MeshStreamRequest, MeshTaskRequest, MeshTaskResult, MeshTopology,
    MeshTopologyRequest,
};
use crate::json::Value;
use crate::result::Outcome;
use crate::session::SessionRef;

/// The kind of work a module performs.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum ModuleType {
    /// Gathers facts about a target without changing it.
    Survey,
    /// Performs an offensive action that may open a session.
    Exploit,
    /// Generates payloads for delivery by other modules.
    PayloadProvider,
}

impl ModuleType {
    /// The wire string for this type.
    pub fn as_str(self) -> &'static str {
        match self {
            ModuleType::Survey => "survey",
            ModuleType::Exploit => "exploit",
            ModuleType::PayloadProvider => "payload_provider",
        }
    }
}

/// A single configuration field a module needs.
#[derive(Clone, Debug)]
pub struct Requirement {
    pub key: String,
    pub value_type: String,
    pub required: bool,
    pub default: String,
    pub description: String,
    pub allowed: Vec<String>,
    pub secret: bool,
}

impl Requirement {
    /// Builds a required requirement of the given type.
    pub fn new(key: &str, value_type: &str, description: &str) -> Requirement {
        Requirement {
            key: key.to_string(),
            value_type: value_type.to_string(),
            required: true,
            default: String::new(),
            description: description.to_string(),
            allowed: Vec::new(),
            secret: false,
        }
    }

    pub(crate) fn to_value(&self) -> Value {
        Value::object(vec![
            ("key", Value::Str(self.key.clone())),
            ("type", Value::Str(self.value_type.clone())),
            ("required", Value::Bool(self.required)),
            ("default", Value::Str(self.default.clone())),
            ("description", Value::Str(self.description.clone())),
            (
                "allowed",
                Value::Array(self.allowed.iter().cloned().map(Value::Str).collect()),
            ),
            ("secret", Value::Bool(self.secret)),
        ])
    }
}

/// The metadata a module reports during the handshake. `name`, `version`, and
/// `module_type` are required; Hovel treats this handshake as authoritative
/// over any package-manifest hints.
pub struct Info {
    pub name: String,
    pub version: String,
    pub module_type: ModuleType,
    pub summary: String,
    pub description: String,
    pub tags: Vec<String>,
    pub discovery_context: Vec<(String, Value)>,
}

/// The configuration contract a module reports.
#[derive(Default)]
pub struct Schema {
    pub chain_config: Vec<Requirement>,
    pub target_config: Vec<Requirement>,
    pub outputs: Vec<(String, Value)>,
    pub planning_context: Vec<(String, Value)>,
}

/// Implemented by every Hovel module. `info` and `schema` must be cheap and
/// side-effect free; `run` does the work when the module is thrown.
pub trait Module {
    fn info(&self) -> Info;
    fn schema(&self) -> Schema;
    fn run(&self, ctx: &mut Context) -> Outcome;

    fn describe_mesh(
        &self,
        _req: MeshDescribeRequest,
    ) -> Result<MeshDescriptor, String> {
        Err(format!("module {:?} is not a mesh provider", self.info().name))
    }

    fn mesh_topology(
        &self,
        _req: MeshTopologyRequest,
    ) -> Result<MeshTopology, String> {
        Err(format!(
            "module {:?} does not implement mesh.topology",
            self.info().name
        ))
    }

    fn list_mesh_beacons(
        &self,
        _req: MeshBeaconRequest,
    ) -> Result<Vec<MeshBeacon>, String> {
        Err(format!(
            "module {:?} does not implement mesh.beacons",
            self.info().name
        ))
    }

    fn run_mesh_task(
        &self,
        _ctx: &mut Context,
        _req: MeshTaskRequest,
    ) -> Result<MeshTaskResult, String> {
        Err(format!(
            "module {:?} does not implement mesh.task",
            self.info().name
        ))
    }

    fn open_mesh_stream(
        &self,
        _ctx: &mut Context,
        _req: MeshStreamRequest,
    ) -> Result<SessionRef, String> {
        Err(format!(
            "module {:?} does not implement mesh.open_stream",
            self.info().name
        ))
    }
}
